xtask/src/main.rs

```diff
@@ -0,0 +1,328 @@
+use std::{
+    io::{BufRead, BufReader},
+    path::Path,
+    process::{ExitCode, Stdio},
+};
+
+use anyhow::{bail, ensure, Context};
+use cli::Args;
+
+use crate::{
+    cli::{Subcommand, ValidateHlslCommand, ValidateSubcommand},
+    fs::{open_file, remove_dir_all, remove_file},
+    glob::visit_files,
+    path::join_path,
+    process::{which, EasyCommand},
+    result::LogIfError,
+};
+
+mod cli;
+mod fs;
+mod glob;
+mod path;
+mod process;
+mod result;
+
+fn main() -> ExitCode {
+    env_logger::builder()
+        .filter_level(log::LevelFilter::Info)
+        .parse_default_env()
+        .format_indent(Some(0))
+        .init();
+
+    let args = Args::parse();
+
+    match run(args) {
+        Ok(()) => ExitCode::SUCCESS,
+        Err(e) => {
+            log::error!("{e:?}");
+            ExitCode::FAILURE
+        }
+    }
+}
+
+fn run(args: Args) -> anyhow::Result<()> {
+    let snapshots_base_out = join_path(["tests", "out"]);
+
+    let Args { subcommand } = args;
+
+    assert!(which("cargo").is_ok());
+
+    match subcommand {
+        Subcommand::All => {
+            EasyCommand::simple("cargo", ["fmt"]).success()?;
+            EasyCommand::simple("cargo", ["test", "--all-features", "--workspace"]).success()?;
+            EasyCommand::simple(
+                "cargo",
+                [
+                    "clippy",
+                    "--all-features",
+                    "--workspace",
+                    "--",
+                    "-D",
+                    "warnings",
+                ],
+            )
+            .success()?;
+            Ok(())
+        }
+        Subcommand::Clean => {
+            let mut found_err = false;
+            visit_files(
+                ".",
+                "*.{metal,air,metallib,vert,frag,comp,spv}",
+                &mut found_err,
+                |file, _| {
+                    remove_file(file).with_context(|| format!("failed to remove file {file:?}"))
+                },
+            );
+            ensure!(
+                !found_err,
+                "failed to clean one or more files, see above output for more details"
+            );
+            Ok(())
+        }
+        Subcommand::Bench { clean } => {
+            if clean {
+                let criterion_artifact_dir = join_path(["target", "criterion"]);
+                log::info!("removing {}", criterion_artifact_dir.display());
+                remove_dir_all(&criterion_artifact_dir)
+                    .with_context(|| format!("failed to remove {criterion_artifact_dir:?}"))?;
+            }
+            EasyCommand::simple("cargo", ["bench"]).success()
+        }
+        Subcommand::Validate(cmd) => {
+            let mut found_err = false;
+            let ack_visiting = |path: &Path| log::info!("Validating {}", path.display());
+            match cmd {
+                ValidateSubcommand::Spirv => {
+                    visit_files(
```

Comment:
```suggestion
                    let spirv_as = "spirv-as";
                    let spirv_val = "spirv-val";
                    which(spirv_as)?;
                    which(spirv_val)?;
                    visit_files(
```

Reply:
D'oh, good catch! Implemented a fix with 3b31bab5 that's very similar to yours.